### PR TITLE
Add dev documentation

### PR DESCRIPTION
**Pull request by JensMueller2709**
- [x] Rework existing documentation
- [x] Transformation logic
- [x] Supported Node types
- [x]  Write ADR for some decisions
- [x]  How to add new node types
- [x]  Known Issues
- [x]  Adapt mkdocs

### CODE REVIEW

File: docs/dev/plugins/cloudFoundry/developer_guide/Script-Overview.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# CF-Plugin Scripts overview
+To transform without pushing anything we need to extract some logic out of the transformation and add it to deploy scripts.<br/>
+All python scripts you will find in the resources see [here](https://github.com/StuPro-TOSCAna/TOSCAna/tree/master/server/src/main/resources/cloudFoundry/deployment_scripts).<br/>
+We add all commands and execution of python scripts into the `deploy_application.sh`. This should be done while the transformation process.<br/>
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/developer_guide/Script-Overview.md]: Please dont use useless html tags, just add an empty line between the sentences

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 2]
```diff
@@ -1,78 +1,31 @@
 # Cloud Foundry Plugin
 
 ## Basic information of Cloud Foundry (CF)
-CF is a open-source Platform as a Service model (PaaS) which allows everyone to easily deploy an existing application on an infrastructure e.g. AWS, vSphere or OpenStack.
+CF is a open-source Platform as a Service model (PaaS) which allows everyone to deploy easily an existing application on an infrastructure e.g. AWS, vSphere or OpenStack.
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: Reuse the old one.

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 3]
```diff
@@ -1,78 +1,31 @@
...
+For further information have a look in the transformation documentation see [here](TODO: insert Link).
 
+## Node Types
+The plugin supports some normative node types of the TOSCA specification.
+A developer could adapt the code to provide additional node types.
+A list with all supported node types and how to add new node types, you find in the transformation documentation.
 
 ### Requirements/Issues
-For this transformation-workflow there are some requirements and possible issues:
+The Cloud Foundry has some requirements:
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: Does CF have these requirements or the CF plugin?

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 4]
```diff
@@ -1,78 +1,31 @@
...
-The script has to enable the docker-support. Afterwards the applications can be pushed with the manifest-file as attribute.
-<pre><code>
-...
-cf enable-feature-flag diego_docker
-cf push myAppName manifest.yml
-...
-</code></pre>
+## Transformation
+The plugin transforms the given TOSCA-model into CloudFoundry deployment artifacts.
+There will be created some scripts as well to provide an easy deployment for the user. Therefore the user needs a Cloud Foundry CLI which is connected to a CloudFoundry instance.
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: Reword the first part of the sentence.

Maybe:
> To simplify the deployment of a application, the CF plugin will add several deployment scripts to the Target artifact.

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 5]
```diff
@@ -1,78 +1,31 @@
...
+For further information have a look in the transformation documentation see [here](TODO: insert Link).
 
+## Node Types
+The plugin supports some normative node types of the TOSCA specification.
+A developer could adapt the code to provide additional node types.
+A list with all supported node types and how to add new node types, you find in the transformation documentation.
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: Please replace this with a link once ready

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 6]
```diff
@@ -1,78 +1,31 @@
...
 
 ### Requirements/Issues
-For this transformation-workflow there are some requirements and possible issues:
+The Cloud Foundry has some requirements:
 - using CloudFoundry-Platform assumes that is already installed on the target infrastructure
 - the CloudFoundry-CLI has to be installed on the local machine
+- python has to be installed
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: Python 2 or 3?

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 7]
```diff
@@ -1,78 +1,31 @@
...
+- databases needs a environment variable with name `database_host`. This is a convention which we discusses with our supervisors.
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: Maybe reword:
> Databases need a environment variable with the name `database_host`. We decided this convention with our supervisors.

File: docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md [hunk 8]
```diff
@@ -1,78 +1,31 @@
...
 
 ## Deployment
-If the user wants to deploy the application automatically, the system could simply execute the created deployment-script.
-
+To deploy an application the user just execute the deploy script.
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/CloudFoundryPlugin_Basics.md]: How? Doubleclick? Push the start button on the washing machine?

File: docs/dev/plugins/cloudFoundry/developer_guide/Script-Overview.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# CF-Plugin Scripts overview
+To transform without pushing anything we need to extract some logic out of the transformation and add it to deploy scripts.<br/>
+All python scripts you will find in the resources see [here](https://github.com/StuPro-TOSCAna/TOSCAna/tree/master/server/src/main/resources/cloudFoundry/deployment_scripts).<br/>
```
Comment (c-mueller) [inline docs/dev/plugins/cloudFoundry/developer_guide/Script-Overview.md]: would call `resources` -> `resource folder`

File: docs/dev/plugins/cloudFoundry/Python_Scripts_Overview.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+ }
+```
+The python script reads this file and knows the meaning of each variable. The placeholder names are fixed values in the java code and in the python script as well. It is important that these values are the same.   
+Afterwards the script is able to read the credentials of the service and set the values to the environment variable.   
+Also this script creates a config file with all service credentials for later use.   
+Lamp Example: `cf set-env $appName database_user = $database_user`
+
+## Python replace paths in file
+In some scripts could be paths which are not available on the warden container. So this paths has to be replaced with existing paths.   
+Twelve Factor: There are no paths which depends on the environment.
```
Comment (orferch) [inline docs/dev/plugins/cloudFoundry/Python_Scripts_Overview.md]: depend

File: docs/dev/plugins/cloudFoundry/Python_Scripts_Overview.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+## Python configureMysql
+The readCredentials script creates a config file with all needed service credentials which are necessary to create a connection.
+The configureMysql script creates a connection to the service with this config file. If there is a valid connection the script will execute the `.sql` file on the database.   
+A requirement is that there is a `.sql` configure script. An alternative solution is to insert the database init script into the application code.   
+Needs a additional python mysql-python package `mysql.connector`.
+Lamp Example: `mysqlCursor.execute(dbInitCommand)`
+Twelve Factor: there have to be a init database script.
```
Comment (orferch) [inline docs/dev/plugins/cloudFoundry/Python_Scripts_Overview.md]: an init

File: docs/dev/plugins/cloudFoundry/Supported_Node_Types.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+The visitor just adds all files and environment variables to the cf-app. There is nothing more to do because CloudFoundry creates automatically a `MYSQL DatabaseManagementSystem` when creating a service.
+
+### `Apache`
+Nothing to do, because the CloudFoundry instance creates automatically a apache WebServer if needed.
+
+### `WebApplication`
+The name of the WebApplication node will be the name of the cf-app because it is a top node and contains probably the main application logic.   
+Also the visitor checks if there are configure or create scripts existing. If yes the plugin adds the information to the cf-app. These scripts will be executed by a python script during the deployment. To get further information about the scripts have a look [here](Python_Scripts_Overview.md).   
+All files and environment variables will be added to the cf-app as well.   
+Here you see a example of how the visitor is getting the configure script of a node:
```
Comment (orferch) [inline docs/dev/plugins/cloudFoundry/Supported_Node_Types.md]: an example

File: docs/dev/plugins/cloudFoundry/Transformation_Process.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+- per application there is one CloudFoundry application object (cf-app) which contains all information of the application
+- the output is called target artifact
+
+## Visitors
+### Node visitor
+The node visitor contains the logic for the supported node types. It fills the cf-app with all information out of the node.
+
+### Prepare visitor
+The prepare visitor prepares the model before the transformation. It sets some fixed placeholder to properties of some node types (e.g. Mysql-Database). This step is needed to enable the environment recognition.   
+If a property of node points to a property to a node where the properties are replaced by placeholders, the plugin knows afterwards what the pointing variable want to get. The python scripts knows the placeholders as well and can check which environment variables have these values.
```
Comment (orferch) [inline docs/dev/plugins/cloudFoundry/Transformation_Process.md]: wants to get

File: docs/dev/plugins/cloudFoundry/Transformation_Process.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+- A service (node type which the plugin handles as service e.g. MYSQL database node) gets its own stack, because every service is a top node (only connectsTo relations as incoming edges).
+- We share this logic with the Kubernetes plugin. So we are using the same methods to do this.
+
+### Getting necessary information (`transform()`)
+In this step the plugin goes through every stack of nodes and calls the node visitor. The node visitor collects all necessary information like inputs, disk-size, paths and so on from the node and add the information to the cf-app.   
+If there is a node type which is declared as service, the plugin checks the source node/s of the connects-to connection and adds the service with all belonging information to the source application/s. The already created cf-app of the stack from the service node, will be deleted.   
+After this step all information of the nodes are in the cf-apps.
+
+## File Creator
+The File Creator is one of the most important class in the Cloud Foundry plugin. This class creates all files for the target artifact.
```
Comment (orferch) [inline docs/dev/plugins/cloudFoundry/Transformation_Process.md]: most important classes

File: docs/dev/plugins/cloudFoundry/Transformation_Process.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    database_port: 3306
+    database_password: cf_database_password_placeholder_my_db
+  services:
+    - my_db
+```
+
+### Services
+If the cf-apps contains a service the FileCreator looks for a suitable service of the provide. Therefore the plugin creates a connection to the provider with the given credentials (with credentials given from the user). It receives all services which are offered of the provider. To do that the plugin implements a (java cf client)[https://github.com/cloudfoundry/cf-java-client].    
+A CloudFoundry service consists of a service name and a service plan. The service name can differ from provider to provider, that's the cause why the plugin needs a connection to the CloudFoundry provider. The service plan specifies the settings of the services e.g. size of a database service. The plugin is looking for a key word (for a mysql service the keyword is `mysql`) in the service descriptions and chooses a suitable service name. Also it chooses a free plan as default.   
+All services will be printed in a extra file called `all_services.txt` in the output folder, so that the user could select another service or change the plan.
```
Comment (orferch) [inline docs/dev/plugins/cloudFoundry/Transformation_Process.md]: an extra

File: docs/dev/adr/0019-cloudfoundry-deploymentscripts.md [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+## Decision Outcome
+
+* Chosen Alternative: python scripts
+
+## Pros and Cons of the Alternatives
+
+### Setting stuff manually by user
+
+* `+` provides many options for user
```
Comment (mmunozba) [inline docs/dev/adr/0019-cloudfoundry-deploymentscripts.md]: add an article: `... for user`

File: docs/dev/plugins/cloudFoundry/Basics.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
+# Cloud Foundry Plugin
+
+## Basic information of Cloud Foundry (CF)
+CF is a open-source Platform as a Service model (PaaS) which allows everyone to easily deploy an existing application on an infrastructure e.g. AWS, vSphere or OpenStack.   
+CF installs everything what the application needs to compile and run like OS, buildpack with all languages, libraries, and services that the app uses.
```
Comment (mmunozba) [inline docs/dev/plugins/cloudFoundry/Basics.md]: Change `...everything what the application needs...` to `...everything that the application needs...`

File: docs/dev/plugins/cloudFoundry/Known_challenges.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
+# Known Issues and Requirements in the CloudFoundry plugin
+The purpose of this document is to collect all known issues in the CloudFoundry plugin. (This document is not exhaustive)
+
+## Conventions
+- If using a database the existing of an environment variable named `database_host`
+- For each transformation a valid connection to a CloudFoundry instance is needed to get information about available services
+- All environment variables will be set globally
+
+## General problems
+- Most of the user provided data cannot be accounted because CloudFoundry will create automatically their own data.
```
Comment (mmunozba) [inline docs/dev/plugins/cloudFoundry/Known_challenges.md]: Add a hyphen: `user provided` -> `user-provided`

File: docs/dev/plugins/cloudFoundry/Supported_Node_Types.md [hunk 19]
```diff
@@ -0,0 +1,90 @@
+# Cloud Foundry - supported NodeTypes
+
+## Overview
+|Node Type | Doing | Description |Implemented |
```
Comment (mmunozba) [inline docs/dev/plugins/cloudFoundry/Supported_Node_Types.md]: What is `Doing` supposed to mean?

File: docs/dev/plugins/cloudFoundry/Transformation_Process.md [hunk 20]
```diff
@@ -0,0 +1,151 @@
+# Cloud Foundry - Transformation Process
+
+The purpose of this document is to show the transformation logic and process of the CloudFoundry plugin for a developer.
+This document does not mention all classes which are used in the java code but describes the method of the transformation so that the developer is able to understand the code and the transformation process.   
+This is the text based documentation, you can find an example transformation with some pictures [here](Transformation_example.pdf).
+
+## Foreword
+- the TOSCA model is a graph. So we are talking about some graph terms. For further information please read the documentation of the core/model.
```
Comment (mmunozba) [inline docs/dev/plugins/cloudFoundry/Transformation_Process.md]: `So we are talking about some graph terms` sounds a bit too informal.

File: docs/dev/plugins/cloudFoundry/Transformation_Process.md [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+- plugin gets some properties (at the beginning of transformation) from the user to create connection to CloudFoundry instance.
+- plugin is using visitor patterns for nodes. The visitors contain the logic for the node types.
+- there will be created one manifest for all applications in the model
+- per application there is one CloudFoundry application object (cf-app) which contains all information of the application
+- the output is called target artifact
+
+## Visitors
+### Node visitor
+The node visitor contains the logic for the supported node types. It fills the cf-app with all information out of the node.
```
Comment (mmunozba) [inline docs/dev/plugins/cloudFoundry/Transformation_Process.md]: Consider using the actual class names `NodeVisitor` and `PrepareVisitor` at some points for better understanding.

File: mkdocs.yml [hunk 22]
```diff
@@ -39,9 +39,24 @@ pages:
             - "0009 - CloudFoundry CLI": "dev/adr/0009-cloudfoundry-cli.md"
             - "0010 - How to perform Kubernetes end to end tests": "dev/adr/0010-use-docker-java.md"
             - "0011 - Put Integration Tests in seperate source directory": "dev/adr/0011-put-integration-tests-in-seperate-folder.md"
-            - "0012 - Use java CloudFoundry Library": "dev/adr/0012-cf-java-library.md"
+            - "0012 - CloudFoundry java Library": "dev/adr/0012-cloudfoundry-java-library.md"
```
Comment (mmunozba) [inline mkdocs.yml]: Capitalise the `Java` in the title

File: docs/dev/plugins/cloudFoundry/Transformation_Process.md [hunk 23]
```diff
@@ -0,0 +1,151 @@
+# Cloud Foundry - Transformation Process
+
+The purpose of this document is to show the transformation logic and process of the CloudFoundry plugin for a developer.
+This document does not mention all classes which are used in the java code but describes the method of the transformation so that the developer is able to understand the code and the transformation process.   
+This is the text based documentation, you can find an example transformation with some pictures [here](Transformation_example.pdf).
```
Comment (mmunozba) [inline docs/dev/plugins/cloudFoundry/Transformation_Process.md]: Why use a `.pdf` file here? 😮
